compiler/typecheck/TcType.hs

```diff
@@ -1815,26 +1815,26 @@ tc_eq_type view_fun orig_ty1 orig_ty2 = go True orig_env orig_ty1 orig_ty2
 
     orig_env = mkRnEnv2 $ mkInScopeSet $ tyCoVarsOfTypes [orig_ty1, orig_ty2]
 
-    -- @eqFunTy arg res ty@ is True when @ty@ equals @FunTy arg res@. This is
+    -- @eqFunTy w arg res ty@ is True when @ty@ equals @FunTy w arg res@. This is
     -- sometimes hard to know directly because @ty@ might have some casts
     -- obscuring the FunTy. And 'splitAppTy' is difficult because we can't
     -- always extract a RuntimeRep (see Note [xyz]) if the kind of the arg or
     -- res is unzonked/unflattened. Thus this function, which handles this
     -- corner case.
-    eqFunTy :: Bool -> RnEnv2 -> Type -> Type -> Type -> Maybe Bool
-    eqFunTy vis env arg res (FunTy _ arg' res')
-      = go vis env arg arg' <!> go vis env res res'
-    eqFunTy vis env arg res ty@(AppTy{})
-      | Just (tc, [_, _, arg', res']) <- get_args ty []
+    eqFunTy :: Bool -> RnEnv2 -> Mult -> Type -> Type -> Type -> Maybe Bool
+    eqFunTy vis env w arg res (FunTy w' arg' res')
+      = go vis env (fromMult w) (fromMult w') <!> go vis env arg arg' <!> go vis env res res'
+    eqFunTy vis env w arg res ty@(AppTy{})
+      | Just (tc, [w', _, _, arg', res']) <- get_args ty []
```

Comment:
Ouch!

Did we see a bug that this triggered?

Reply:
I didn't find any bug caused by this.